### PR TITLE
OpenTracing support for grpc-java

### PR DESCRIPTION
**Pull request by kcamenzind**
This PR provides support for OpenTracing in gRPC-java clients and servers. 

**Purpose**
Distributed tracing has become increasingly important to gain visibility into the application stack, since distributed systems (such as ones that use gRPC) propagate requests across process boundaries. OpenTracing is a set of consistent, expressive, vendor-neutral APIs for distributed tracing and context propagation, and by providing support for OpenTracing in gRPC, we make it easy for developers to trace their gRPC requests.

**Usage**
This package allows gRPC-java users to trace client calls and server responses by adding a `ClientTracingInterceptor` and `ServerTracingInterceptor` to their clients and servers respectively. We also utilize the `io.grpc.Context` to allow users to gain access to the current span during server-side request handling, in order to continue the trace further into the application stack.

In order to actually produce a trace, developers can bind OpenTracing to any tracing implementation that uses a span model (such as Uber's [Jaeger Bindings](https://github.com/uber/jaeger-client-java), an open-source Zipkin-compatible tracer). 

**Next Steps**

Any comments on this implementation are welcome. We'd also like to know whether gRPC wants to host these artifacts, or whether we should publish them under `io.opentracing.contrib.grpc`.

### CODE REVIEW

**Comment by bhs** [comment]
cc @zinuga: how should we proceed with this? We don't have commit rights to the repo... Thanks.

**Comment by zinuga** [comment]
@bensigelman and @kcamenzind I have invited you to join grpc-opntracing-team and the team has admin writes to this repo.

File: java/README.rst [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+This package is available on Maven Central and can be added to your project as follows:
+
+**Maven**
+
+.. code-block::
+
+    <dependencies>
+        <dependency>
+            <groupId>io.opentracing.contrib.grpc</groupId>
```
Comment (kcamenzind) [inline java/README.rst]: drop the .grpc

File: java/README.rst [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        .withOperationName(new OperationNameConstructor() {
+            @Override
+            public <ReqT, RespT> String constructOperationName(MethodDescriptor<ReqT, RespT> method) {
+                // construct some operation name from the method descriptor
+            }
+        })
+        .withActiveSpanSource(new ActiveSpanSource() {
+            @Override
+            public Span getActiveSpan() {
+                // implement how to get the current active span
```
Comment (kcamenzind) [inline java/README.rst]: maybe point at one of the easier/prefab things in this comment?

File: java/src/generated/io/opentracing/contrib/grpc/GreeterGrpc.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
+package io.opentracing.contrib.grpc;
```
Comment (kcamenzind) [inline java/src/generated/io/opentracing/contrib/grpc/GreeterGrpc.java]: rename this path `s/generated/testgen/` or `testgenfiles` or???

File: java/src/main/java/io/opentracing/contrib/grpc/ActiveSpanSource.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            return null;
+        }
+    };
+    
+    /**
+     * ActiveSpanSource implementation that returns the
+     *  current span stored in the GRPC context under
+     *  {@link OpenTracingContextKey}
+     */
+    public static ActiveSpanSource GRPC = new ActiveSpanSource() {
```
Comment (kcamenzind) [inline java/src/main/java/io/opentracing/contrib/grpc/ActiveSpanSource.java]: `GRPC_CONTEXT`?

File: java/src/main/java/io/opentracing/contrib/grpc/ClientRequestAttribute.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package io.opentracing.contrib.grpc;
+
+/**
+ * List of client request attributes available to be traced.
+ */
+public enum ClientRequestAttribute {
```
Comment (kcamenzind) [inline java/src/main/java/io/opentracing/contrib/grpc/ClientRequestAttribute.java]: nit: make this a public enum inside CTInt.java?

File: java/src/main/java/io/opentracing/contrib/grpc/OpenTracingContextKey.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * A {@link io.grpc.Context} key for the current OpenTracing trace state.
+ * 
+ * Can be used to get the active span, or to set the active span for a scoped unit of work. 
+ * See the <a href="../../../../../../README.rst">grpc-java OpenTracing docs</a> for use cases and examples.
+ */
+public class OpenTracingContextKey {
+    
+    public static final String KEY_NAME = "ot-active-span";
```
Comment (kcamenzind) [inline java/src/main/java/io/opentracing/contrib/grpc/OpenTracingContextKey.java]: let's read the docs on the convention here

File: java/src/main/java/io/opentracing/contrib/grpc/ServerRequestAttribute.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package io.opentracing.contrib.grpc;
+
+/**
+ * List of server request attributes available to be traced.
+ */
+public enum ServerRequestAttribute {
```
Comment (kcamenzind) [inline java/src/main/java/io/opentracing/contrib/grpc/ServerRequestAttribute.java]: (same as above)

**Comment by bhs** [comment]
Looks excellent! Merging. Thanks again for all of your work on this.

**Pull request merged**